packages/file-search/src/node/file-search-service-impl.ts

```diff
@@ -0,0 +1,106 @@
+import { FileUri } from '@theia/core/lib/node';
+import * as fuzzy from 'fuzzy';
+import { injectable, inject } from 'inversify';
+import * as parser from 'gitignore-parser';
+import { ILogger } from '@theia/core';
+
+type Filtered = (simpleName: string, fullPath: string) => boolean;
+
+@injectable()
+export class FileSearchServiceImpl implements FileSearchService {
+
+    constructor( @inject(ILogger) private logger: ILogger) { }
+
+    async find(uri: string, searchPattern: string, options?: { fuzzyMatch?: boolean, limit?: number, useGitignore?: boolean }): Promise<string[]> {
+        const basePath = FileUri.fsPath(new URI(uri));
+        const opts = {
+            fuzzyMatch: true,
+            limit: Number.MAX_SAFE_INTEGER,
+            useGitignore: true,
+            ...options
+        };
+        const result: string[] = [];
+        const limitReached = new Error("limit reached");
+        const filtered = this.getFiltered(basePath, opts);
+        try {
+            this.findRecursive(basePath, filtered, filePath => {
+                if (result.length >= opts.limit) {
+                    throw limitReached;
+                }
+                if (opts.fuzzyMatch && fuzzy.test(searchPattern, filePath)) {
+                    result.push(FileUri.create(filePath).toString());
+                } else {
+                    if (filePath.toLocaleLowerCase().indexOf(searchPattern.toLocaleLowerCase()) !== -1) {
+                        result.push(FileUri.create(filePath).toString());
+                    }
+                }
+            });
+        } catch (e) {
+            if (e !== limitReached) {
+                throw e;
+            }
+        }
+        return result;
+    }
+
+    private findRecursive(filePath: string, filtered: Filtered, acceptor: (fileName: string) => void) {
+        const result = fs.readdirSync(filePath);
+        for (const child of result) {
+            const childPath = path.join(filePath, child);
+            if (!filtered(child, childPath)) {
+                if (fs.statSync(childPath).isDirectory()) {
+                    this.findRecursive(childPath, filtered, acceptor);
+                } else {
+                    acceptor(childPath);
+                }
+            }
+        }
+    }
+
+    private getFiltered(basePath: string, options: { fuzzyMatch: boolean, limit: number, useGitignore: boolean }): Filtered {
+        const defaultFilter = (simpleName: string) =>
+            simpleName === '.git' ||
+            simpleName.startsWith('.git/');
+        if (options.useGitignore) {
+            const gitIgnore = this.findGitIgnore(basePath);
+            if (gitIgnore) {
+                const matcher = parser.compile(gitIgnore.contents);
+                return (short, long) => defaultFilter(short) || matcher.denies(short);
+            }
+        }
+        return defaultFilter;
+    }
+
+    private findGitIgnore(basePath: string): { path: string, contents: string } | undefined {
```

Comment:
This doesn't take into account multiple git repos inside the workspace.
However the git extension supports this. So maybe we should support it here too ?

It would be nice to know if we have the git extension and if so use it to get the possible git repos and use the .gitignores appropriately.

I'm curious if such a method (having kinda of an optional extension) is possible too since I may have other use cases for it.

Reply:
CC: @kittaakos

Reply:
Yes, we should respect nested git repos. Actually, we should also respect nested `.gitignore` files. I don't think there is a value in asking the git extensions about that, though.